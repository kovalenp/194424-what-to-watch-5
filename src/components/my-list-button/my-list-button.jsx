import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {movieProps} from "../../validation/propTypes";
import {setFavorite, removeFavorite, pullComments, getCurrentMovie, getPromoMovie} from "../../services/movie-service";
import {AppRoute} from "../../common/constants";
import browserHistory from "../../common/browser-history";

const MyListButton = (props) => {

  const {movie, isAuth, promo} = props;

  const isFavorite = () => movie.is_favorite;

  const getButtonSvg = () => {
    if (isFavorite()) {
      return (
        <svg viewBox="0 0 18 14" width="18" height="14">
          <use xlinkHref="#in-list"></use>
        </svg>
      );
    } else {
      return (
        <svg viewBox="0 0 19 20" width="19" height="20">
          <use xlinkHref="#add"></use>
        </svg>
      );
    }
  };

  const updateMovieState = (id) => {
    if (promo && promo.id === movie.id) {
      return props.getMovie(id).then(() => props.getPromo());
    } else {
      return props.getMovie(id);
    }
  };

  const performFavoriteAction = (id) => {
    if (!isFavorite()) {
      return setFavorite(id).then(() => updateMovieState(id));
    } else {
      return removeFavorite(id).then(() => updateMovieState(id));
    }
  };

  return (
    <button
      className="btn btn--list movie-card__button"
      type="button"
      onClick={() => {
        if (!isAuth) {
          return browserHistory.push(AppRoute.LOGIN);
        }
        return performFavoriteAction(movie.id);
      }}
    >
      {getButtonSvg()}
      <span>My list</span>
    </button>
  );
};

MyListButton.propTypes = {
  movie: movieProps,
  promo: movieProps,
  isAuth: PropTypes.bool,
  getMovie: PropTypes.func,
  getPromo: PropTypes.func,
};

const mapStateToProps = (state) => {
  return {
    promo: state.MOVIES.promo
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    getComments: (movieId) => dispatch(pullComments(movieId)),
    getMovie: (id) => dispatch(getCurrentMovie(id)),
    getPromo: () => dispatch(getPromoMovie())
  };
};

export {MyListButton};
export default connect(mapStateToProps, mapDispatchToProps)(MyListButton);
