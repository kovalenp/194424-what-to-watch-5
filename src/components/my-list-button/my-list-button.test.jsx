import React from "react";
import renderer from "react-test-renderer";
import {MyListButton} from "./my-list-button";
import movie from "../../mocks/movie";

describe(`MyListButton meets B21`, () => {

  it(`and rendered correctly for auth'ed`, () => {
    const tree = renderer.create((
      <MyListButton
        movie={movie}
        promo={movie}
        getMovie={() => movie}
        getPromo={()=>movie}
        isAuth={true} />
    )).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`and rendered correctly for non auth'ed`, () => {
    const tree = renderer.create((
      <MyListButton
        movie={movie}
        promo={movie}
        getMovie={() => movie}
        getPromo={()=>movie}
        isAuth={false} />
    )).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
