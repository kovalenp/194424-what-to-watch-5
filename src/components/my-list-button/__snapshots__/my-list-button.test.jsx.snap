// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyListButton meets B21 and rendered correctly for auth'ed 1`] = `
<button
  className="btn btn--list movie-card__button"
  onClick={[Function]}
  type="button"
>
  <svg
    height="14"
    viewBox="0 0 18 14"
    width="18"
  >
    <use
      xlinkHref="#in-list"
    />
  </svg>
  <span>
    My list
  </span>
</button>
`;

exports[`MyListButton meets B21 and rendered correctly for non auth'ed 1`] = `
<button
  className="btn btn--list movie-card__button"
  onClick={[Function]}
  type="button"
>
  <svg
    height="14"
    viewBox="0 0 18 14"
    width="18"
  >
    <use
      xlinkHref="#in-list"
    />
  </svg>
  <span>
    My list
  </span>
</button>
`;
